### ISSUE TITLE
Exits after decoding a few seconds

### ISSUE THREAD

**Issue by barf** [fbieberly/ORBCOMM-receiver#5]
On a raspberry pi I have trouble receiving more than a few seconds of data.
```
$ python3 realtime_receiver.py 
Detached kernel driver
Found Rafael Micro R820T tuner
[R82XX] PLL not locked!

Satellites overhead: 
orbcomm fm108       : 3.1 degrees elevation
Receiving from: orbcomm fm108
Satellite frequencies: [137460000.0, 137712500.0]
Recording samples.
Create multiprocessing queue
Creating multiprocessing process
Frequency offset: -225.44031311154598
Bit stream offset: 80
Fill: data: 2DAF89852278C41BA219 
Fill: data: 483E6B7369E174F69B78 
Fill: data: D1C3CE158E877B727689 
Fill: data: EFB9B7EF01F3BD2EC8CB 
Fill: data: 62D0CF10E6C6EB5B4129 
Fill: data: 8F08F3CC94BFC8389881 
Fill: data: 3CCE1007D74EF60AD937 
Fill: data: DAB7BAFDC2BB293DF78C 
Fill: data: 817A13E6CAE0EC9108F1 
Fill: data: 0B10D6CD6BBA652F4C90 
Fill: data: 4916C0AEB77D0064FA79 
Fill: data: 85DFA48C4A2BAF54FC76 
Fill: data: 036428E3816DC5C5A0D5 
Fill: data: 530472DFC6A98E8AC854 
Fill: data: 8C3CBA9958B8185B26BE 
Fill: data: 39FA4B5A83A0FD35DF61 
Fill: data: C2099BBD647969AE4AD7 
Fill: data: EAD481E2AC38D6C65D64 
Fill: data: 3815D1C5F70756E3EE14 
Fill: data: 5BF2CD2AFEBD2B5BDDFC 
Fill: data: 1B3958A56A1A7A9BE0ED 
Fill: data: 19F1D8528E1EF9701DED 
Fill: data: 5A8C1A5E354CE775C6A3 
Ending async RTLSDR processing
Fill: data: 6FE4A5F1DDC8B12CADE5 
Fill: data: 567DC3683187453D728D 
Fill: data: DAB7BAFDC2BB293DF78C 
Fill: data: 817A13E6CAE0EC9108F1 
Fill: data: 0B10D6CD6BBA652F4C90 
Reattached kernel driver
Program exited.
```

The RTL-SDR dongle appears to reset itself while the application is running:
```
May 18 03:47:05 otherpi kernel: [23070.641536] r820t 4-001a: destroying instance
May 18 03:47:05 otherpi kernel: [23070.641755] dvb_usb_v2: 'Realtek RTL2832U reference design:1-1.1.2' successfully deinitialized and disconnected
May 18 03:47:12 otherpi kernel: [23077.199329] usb 1-1.1.2: dvb_usb_v2: found a 'Realtek RTL2832U reference design' in warm state
May 18 03:47:12 otherpi kernel: [23077.253125] usb 1-1.1.2: dvb_usb_v2: will pass the complete MPEG2 transport stream to the software demuxer
May 18 03:47:12 otherpi kernel: [23077.253157] dvbdev: DVB: registering new adapter (Realtek RTL2832U reference design)
May 18 03:47:12 otherpi kernel: [23077.264471] i2c i2c-3: Added multiplexed i2c bus 4
May 18 03:47:12 otherpi kernel: [23077.264480] rtl2832 3-0010: Realtek RTL2832 successfully attached
May 18 03:47:12 otherpi kernel: [23077.264517] usb 1-1.1.2: DVB: registering adapter 0 frontend 0 (Realtek RTL2832 (DVB-T))...
May 18 03:47:12 otherpi kernel: [23077.264654] r820t 4-001a: creating new instance
May 18 03:47:12 otherpi kernel: [23077.271584] r820t 4-001a: Rafael Micro r820t successfully identified
May 18 03:47:12 otherpi kernel: [23077.281059] Registered IR keymap rc-empty
May 18 03:47:12 otherpi kernel: [23077.281159] rc rc0: Realtek RTL2832U reference design as /devices/platform/soc/3f980000.usb/usb1/1-1/1-1.1/1-1.1.2/rc/rc0
May 18 03:47:12 otherpi kernel: [23077.281259] input: Realtek RTL2832U reference design as /devices/platform/soc/3f980000.usb/usb1/1-1/1-1.1/1-1.1.2/rc/rc0/input20
May 18 03:47:12 otherpi kernel: [23077.281516] rc rc0: lirc_dev: driver dvb_usb_rtl28xxu registered at minor = 0, raw IR receiver, no transmitter
May 18 03:47:12 otherpi kernel: [23077.281699] usb 1-1.1.2: dvb_usb_v2: schedule remote query interval to 200 msecs
May 18 03:47:12 otherpi kernel: [23077.289770] usb 1-1.1.2: dvb_usb_v2: 'Realtek RTL2832U reference design' successfully initialized and connected
```
I don't experience any reset like this using other SDR applications such as rtl_fm, rtl_tcp or gqrx etc, and I have tested different (known-good) RTL-SDRv3 dongles.

**Comment by fbieberly** [comment]
I haven't run the code on a raspberry pi, but I think the pi will be too slow (my code is not optimized well) to keep up with the data from the RTL-SDR. The data from the RTL-SDR is put into a queue and when it gets full, the application shuts down. I'm almost positive that's what is happening here.
You can look at [Line #115](https://github.com/fbieberly/ORBCOMM-receiver/blob/master/realtime_receiver.py#L115) of the realtime receiver to see where it is happening.
Unfortunately, there's no easy way to speed up the processing. I'd need to go through and find where I can optimize the DSP. Also, if you haven't already, you can turn off all the plotting in the [CONFIG.py](https://github.com/fbieberly/ORBCOMM-receiver/blob/master/CONFIG.py#L10) file.

**Comment by barf** [comment]
Oh I understand, thanks for looking anyway!

Unfortunately my antenna does not reach to my PC. I will try to move the DSP to my PC using the `RtlSdrTcpServer` and `RtlSdrTcpClient`in pyrtlsdr, or is there an easier method to make it connect to the rtl_tcp server?

**Comment by fbieberly** [comment]
I haven't tried using the RtlSdrTcpServer/Client with this code before. Looks like you could just change [Line #84](https://github.com/fbieberly/ORBCOMM-receiver/blob/master/realtime_receiver.py#L84) of the realtime receiver to: 
```sdr = RtlSdrTcpClient(hostname='192.168.1.100', port=12345)``` (with the appropriate ip/port numbers of course). 

Please let me know if that ends up working. That'd be really nice to know.

Depending on what you want to get out of this code, another option is to do the offline recording and decoding. The pi should be able to handle that pretty easily.

**Comment by barf** [comment]
Thanks. I tried the TCP functionality in pyrtlsdr but it was not happy about asyncio, hopefully they can fix that upstream

```
Traceback (most recent call last):
  File "realtime_receiver.py", line 354, in <module>
    sdr.read_samples_async(rtlsdr_callback, num_samples_per_recording, context_dict)
  File "/home/barf/.local/lib/python3.8/site-packages/rtlsdr/rtlsdrtcp/client.py", line 130, in read_samples_async
    raise NotImplementedError('Async read not available in TCP mode')
NotImplementedError: Async read not available in TCP mode
```

**Comment by fbieberly** [comment]
That's too bad. It might not be too hard to use rtl_tcp to do the same thing. I'll look into that over the next couple days, but I'm not promising anything.

I'm going to close this issue, but I'll post a message and push up any rtl_tcp code that I get working. 
Thanks for checking out the code.

**Comment by fbieberly** [comment]
I made a new branch that has a script for receiving samples from a remote computer. If you'd like to try it out, checkout the rtl_tcp branch (it's poorly named since I realized later that I couldn't use the rtl_tcp program, so I switched and used rtl_sdr + netcat). 
I put instructions for how to get it to run in the [readme](https://github.com/fbieberly/ORBCOMM-receiver/tree/rtl_tcp#real-time-recording-and-decoding-from-remote-rtl-sdr).

If you try it out, let me know how it goes.

**Comment by barf** [comment]
thanks! 

I checked out the rtl_tcp branch but it's saying 'RESETTING' over and over (on the PC):
```
$ python3 realtime_receiver_network.py

Satellites overhead: 
orbcomm fm113       : 2.4 degrees elevation
Receiving from: orbcomm fm113
Satellite frequencies: [137662500.0, 137800000.0]
Create multiprocessing queue
Creating socket receiver process
realtime_receiver_network.py:129: DeprecationWarning: The binary mode of fromstring is deprecated, as it behaves surprisingly on unicode inputs. Use frombuffer instead
  temp_samples = np.fromstring(data, dtype=np.uint8).astype(np.float32).view(np.complex64) - 127.0
Frequency offset: -226.41509433962264
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
RESETTING!
```

the RESETTING error appears after running rtl_sdr on the Pi (is the rtl_tcp server not used here?):
```
$ rtl_sdr -f 137500000 -s 1228800 -g 0 - | netcat -uv 192.168.1.4 5556
Found 1 device(s):
  0:  Realtek, RTL2838UHIDIR, SN: 00000001

Using device 0: Generic RTL2832U OEM
Found Rafael Micro R820T tuner
[R82XX] PLL not locked!
Sampling at 1228800 S/s.
Tuned to 137500000 Hz.
Tuner gain set to automatic.
Reading samples in async mode...
Connection to 192.168.1.4 5556 port [udp/freeciv] succeeded!
```

**Comment by fbieberly** [comment]
Thanks again for checking this out. I had a bug in the way the processing code on the PC was receiving the samples which was causing it to drop a bunch of samples. When the decoder can't find enough packets that pass the checksum it assumes that some samples were dropped so it resets all the internal state of the decoder and tries to start over. Unfortunately, because of the bug, that was happening over and over, hence the RESETTING! messages. 

I pushed a fix for that bug and tested it on couple satellite passes and it works for me. It doesn't work as well as just running the code on a computer using the realtime_receiver.py script. It could be because I use UDP to transfer the samples and maybe some packets are arriving out of order or something like that.

I don't use the rtl_tcp server. I had intended to, but then I realized that I didn't know how to interface with it and I didn't find any examples online. I may revisit it in the future. Otherwise, using netcat seems to be mostly effective.
